packages/core/src/lib/ConnectionManager.ts

```diff
@@ -0,0 +1,336 @@
+    pingKeepAlive: number,
+    relayKeepAlive: number
+  ): void {
+    this.libp2p.connectionManager.addEventListener("peer:connect", (evt) => {
+      this.startKeepAlive(evt.detail.remotePeer, pingKeepAlive, relayKeepAlive);
+    });
+
+    /**
+     * NOTE: Event is not being emitted on closing nor losing a connection.
+     * @see https://github.com/libp2p/js-libp2p/issues/939
+     * @see https://github.com/status-im/js-waku/issues/252
+     *
+     * >This event will be triggered anytime we are disconnected from another peer,
+     * >regardless of the circumstances of that disconnection.
+     * >If we happen to have multiple connections to a peer,
+     * >this event will **only** be triggered when the last connection is closed.
+     * @see https://github.com/libp2p/js-libp2p/blob/bad9e8c0ff58d60a78314077720c82ae331cc55b/doc/API.md?plain=1#L2100
+     */
+    this.libp2p.connectionManager.addEventListener("peer:disconnect", (evt) => {
+      this.stopKeepAlive(evt.detail.remotePeer);
+    });
+  }
+
+  private dialPeersInInterval(
+    maxBootstrapPeersAllowed: number,
+    interval: number
+  ): void {
+    // increase interval after every attempt
+    // -> 1st interval: 10 seconds x 2 = 20 seconds
+    // -> 2nd interval: 10 seconds x 3 = 40 seconds
+    // and so on
+    const newInterval = this.dialAttempt * interval;
+    log(`Dialing next set of discovered peers in ${newInterval} ms`);
+
+    // recursively run this function with increase in time
+    this.dialDiscoveredPeers(maxBootstrapPeersAllowed)
+      .then(() => {
+        setTimeout(() => {
+          this.dialPeersInInterval(maxBootstrapPeersAllowed, interval);
+        }, newInterval);
+        this.dialAttempt++;
+      })
+      .catch((err) => {
+        log("Error dialing discovered peers", err);
+      });
+  }
+
+  private async dialDiscoveredPeers(
+    maxBootstrapPeersAllowed: number
+  ): Promise<void> {
+    const {
+      dialableBootstrapPeers,
+      dialableNonBootstrapPeers,
+      bootstrapConnections,
+    } = await this.fetchUpdatedState();
+
+    // find & dial peers found via `dns-discovery`
+    if (
+      (bootstrapConnections.length < maxBootstrapPeersAllowed &&
+        dialableBootstrapPeers.length > 0) ||
+      this.dialAttempt >= DEFAULT_DIAL_ATTEMPTS_BEFORE_BOOTSTRAP_CONNECTION
+    ) {
+      for (let i = 0; i < maxBootstrapPeersAllowed; i++) {
+        const peer = dialableBootstrapPeers[i];
+        if (!peer) break;
+
+        this.dialPeer(peer)
+          .then(() => log(`Dial successful`))
+          .catch((error) => log(error));
+      }
+    }
+
+    //find & dial peers found via discovery mechanisms other than `dns-discovery`
+    for (const peer of dialableNonBootstrapPeers) {
+      this.dialPeer(peer)
+        .then(() => log(`Dial successful`))
+        .catch((error) => log(error));
+    }
+  }
+
+  private async fetchUpdatedState(): Promise<UpdatedStates> {
+    const allPeers = await this.libp2p.peerStore.all();
+
```

Comment:
calling this method set some lock on `peerStore`.

The current solution set an interval at which is will try to dial peers.

The issues is that it can create artificial delay between the moment a peer is discovered and the moment the peer is dialled.

E.g.

- t0 dial attempt
- t2 peer discovered
- t5 second dial attempt

Lost 3 seconds where we could have a done a dial attempt. 3s is very long in terms of web UX.

An event-based approach is more appropriate, especially that the peer store emit a number of events: https://libp2p.github.io/js-libp2p-peer-store/#events

Instead of attempting to dial peers on an interval, attempt to dial peer based on the events of the peer store:
 1. A new peer is discovered, it is added to the peer store
 2. peer store emit an event
 3. connection manager listened to the event, check if the peer needs to be dialled and can be dialled
 4. if yes, then dial peer
 
 With an event based approach, peers are dialled as soon as they are discovered.

Note that in the original issue, I hinted to the usage of event: https://github.com/waku-org/js-waku/issues/914#issuecomment-1372943986

> DNS discovery bootstrap method is used, it discovers and emit nodes that are tagged as bootstrap
> Connection Manager processes the bootstrap node